Tactic tutorial (#513)
leanprover-community/mathlib: docs/extras/tactic_writing.md

particular `tactic unit` is only about manipulating proof state, without
trying to return anything (technically it will return something of type
`unit`, which is the type having exactly one term, denoted by `()`).
Such functions are either called by other
tactics---those are typically in the `tactic` namespace---or called
interactively by users inside tactic blocks---those must be in the
`tactic.interactive` namespace (this is not quite necessary for very simple
tactics, but weird things will happen in general when ignoring this rule). A
shortcut which is sometimes convenient is: one can copy definitions with name
`my_tac1`, `my_tac2`, `my_tac3`, into the `tactic.interactive` namespace using
`` run_cmd add_interactive [`my_tac1,`my_tac2, `my_tac3] ``.
These functions will be used to generate Lean proofs, but we will not prove
anything about these functions themselves, nor will the constants `my_tac1`,
`my_tac2`, etc. appear in the proofs
that they generate. By prefacing them with the keyword `meta`, we tell Lean that
they are for "evaluation purposes only," which disables some of the checks that
non-`meta` declarations must pass.

> This enough knowledge to write a first tactic.

Reviewer: ```suggestion
This is enough knowledge to write a first tactic.
```

example : true :=
begin
  my_first_tactic,
  trivial
end
```

In the example, `my_first_tactic` is underlined in green (in VScode) and
moving the cursor on that line will display our message in the Lean
messages buffer.

Next we need to learn how to chain several actions. Deep down, this is
all about composing functions, but the monadic notations hide this and
emulate imperative programming. We need to use the `and_then`
combinator. The first way is to use the infix notation `>>`, as in:

```lean
meta def my_second_tactic : tactic unit :=
tactic.trace "Hello," >> tactic.trace "World."
```
which now prints our message in two pieces. Alternatively, one can use
the `do` syntax, which has other goodies. This introduces a
comma-separated list of instructions to perform in sequence.
```lean
meta def my_second_tactic' : tactic unit :=
do
  tactic.trace "Hello,",
  tactic.trace "World."
```

Besides displaying messages, the next thing tactics can do is to fail,
potentially with some explanation.
```lean
meta def my_failing_tactic  : tactic unit := tactic.failed

meta def my_failing_tactic' : tactic unit :=
tactic.fail "This tactic failed, we apologize for the inconvenience."
```

> When chaining instructions, the first failure interupts the process.

Reviewer: ```suggestion
When chaining instructions, the first failure interrupts the process.
```

`α` from something of type `tactic α`. It cannot be used to store
regular stuff. The following doesn't work.
```lean
meta def broken_assignment : tactic unit :=
do
 message ← "Hello, World.",  -- WRONG!
 tactic.trace message
```
However, one can use `let` in `do` blocks, as in:
```lean
meta def let_example : tactic unit :=
do
 let message := "Hello, World.",
 tactic.trace message
```
Next, we want to write tactics returning something, as `tactic.target`
is doing. The only extra ingredient is the `return` function. The
following function tries to return `tt` is there is no more goal, `ff`
otherwise. The next one can be used interactively and traces the result
(note that using the first one interactively won't have any visible
effect since interactive use ignores the returned value).
```lean
meta def is_done : tactic bool :=
(tactic.target >> return ff) <|> return tt

meta def trace_is_done : tactic unit :=
is_done >>= tactic.trace
```
The last thing we will need about monadic assignment is pattern-matching
assignment. The following tactic tries to define expressions `l` and `r` as
the left and right hand sides of the current goal. It also uses the

> `to_string` function which is very convenient in combination with `trace` in order to debug tactics, and works on any type which is an instance of `has_to_string`.

Reviewer: ```suggestion
`to_string` function which is very convenient in combination with `trace`
in order to debug tactics, and works on any type which is an instance of `has_to_string`.
```

`tactic expr`, it must be chained with either `>>=` or `←`, as explained
above).
Next `tactic.unify` which, modulo a couple of optional parameters, takes
two expressions and succeeds if and only if they are definitionaly equal.
The first piece of the assumption tactic is a helper function searching
an expression sharing the type of some expression `e` in a list of
expressions, returning the first match (or failing if nothing matches).
```lean
meta def find_matching_type (e : expr) : list expr → tactic expr
| []         := tactic.failed
| (H :: Hs)  := do t ← tactic.infer_type H,
                   (tactic.unify e t >> return H) <|> find_matching_type Hs
```
Make sure you really understand the control flow in the above code using
the previous section. The basic pattern is classical recursive find in a
list. Notice the expression `e` is left of colon, hence it will be passed
unchanged to the recursive call `find_matching_type Hs`. The choice of
name `H` stands for `hypothesis`, while `Hs`, following Haskell's naming
conventions, stands for several hypotheses. The imperative analogue of
what happens for non-empty lists would read something like the following

> impreative pseudo-code

Reviewer: ```suggestion
imperative pseudo-code
```

```
if unify(e, infer_type(H)) then return H else find_matching_type(e, HS)
```
We can now use this function for our interactive tactic. We first need

> to grab the local context using `local_context`, which return a list of

Reviewer: ```suggestion
to grab the local context using `local_context`, which returns a list of
```

new piece we need is `tactic.interactive.«have»` which will create our new
context item. Its weird name works around the fact that `have` is a keyword,
hence not a valid name. It takes two optional arguments that we ignore for
now, and a pre-expression which is a proof of our new item. Such a
pre-expression is constructed using the double-backtick-parenthesis notation:
``` ``(...) ```. Inside such a construction, previously assigned expressions
are accessed using the anti-quotation prefix `%%`. This syntax is very close to the pattern matching syntax we saw above (but different).

```lean
open tactic.interactive («have»)
open tactic (get_local infer_type)

meta def tactic.interactive.add_eq_h₁_h₂ : tactic unit :=
do e1 ← get_local `h₁,
   e2 ← get_local `h₂,
   «have» none none ``(_root_.congr (congr_arg has_add.add %%e1) %%e2)

example (a b j k : ℤ) (h₁ : a = b) (h₂ : j = k) :
  a + j = b + k :=
begin
  add_eq_h₁_h₂,
  exact this
end
```
A last remark about the above tactic: the names `` `h₁ `` and `` `h₂ `` are resolved when the tactic is executed. In order to trigger name resolution when
the tactic is parsed, one should use double-backtick, as in ``` ``h₁ ```. Of course

> 

Reviewer: ```suggestion
```

named `n` in the local context. In our case we will pattern-match on the
parsed location and reject everything except specifying a single name from
the local context. The second new piece is how to parse a user-provided
expression. The relevant parser is `interactive.types.texpr`, whose result is
converted to an actual expression using `tactic.i_to_expr`. This is also the
opportunity for our first serious use of pattern matching assignment, and
for using the second optional argument of `«have»` which is the expected type
(otherwise we would get unapplied multiplication, with an explicit lambda, try it!).
```lean
open interactive (loc.ns)
open interactive.types (texpr location)
meta def mul_left (q : parse texpr) : parse location → tactic unit
| (loc.ns [some h]) := do
   e ← tactic.i_to_expr q,
   H ← get_local h,
   `(%%l = %%r) ← infer_type H,
   «have» h ``(%%e*%%l = %%e*%%r) ``(congr_arg (λ x, %%e*x) %%H),
   tactic.clear H
| _ := tactic.fail "mul_left takes exactly one location"

example (a b c : ℤ) (hyp : a = b) : c*a = c*b :=
begin
  mul_left c at hyp,
  exact hyp
end
```

As a last refinement, let us make a version of this tactic which names the

> multiplied equality by appending `.mul`, and optionaly remove the original

Reviewer: ```suggestion
multiplied equality by appending `.mul`, and optionally removes the original
```
